src/user-event/__tests__/tab.js

```diff
@@ -0,0 +1,329 @@
+import * as userEvent from '..'
+import {setup} from './helpers/utils'
+
+test('should cycle elements in document tab order', async () => {
+  setup(`
+    <div>
+      <input data-testid="element" type="checkbox" />
+      <input data-testid="element" type="radio" />
+      <input data-testid="element" type="number" />
+    </div>`)
+
+  const [checkbox, radio, number] = document.querySelectorAll(
+    '[data-testid="element"]',
+  )
+
+  expect(document.body).toHaveFocus()
+
+  await userEvent.tab()
+
+  expect(checkbox).toHaveFocus()
+
+  await userEvent.tab()
+
+  expect(radio).toHaveFocus()
+
+  await userEvent.tab()
+
+  expect(number).toHaveFocus()
+
+  await userEvent.tab()
+
+  // cycle goes back to first element
+  expect(checkbox).toHaveFocus()
+})
+
+test('should go backwards when shift = true', async () => {
+  setup(`
+    <div>
+      <input data-testid="element" type="checkbox" />
+      <input data-testid="element" type="radio" />
+      <input data-testid="element" type="number" />
+    </div>`)
+
+  const [checkbox, radio, number] = document.querySelectorAll(
+    '[data-testid="element"]',
+  )
+
+  radio.focus()
+
+  await userEvent.tab({shift: true})
+
+  expect(checkbox).toHaveFocus()
+
+  await userEvent.tab({shift: true})
+
+  expect(number).toHaveFocus()
+})
+
+test('should respect tabindex, regardless of dom position', async () => {
+  setup(`
+    <div>
+      <input data-testid="element" tabIndex="2" type="checkbox" />
+      <input data-testid="element" tabIndex="1" type="radio" />
+      <input data-testid="element" tabIndex="3" type="number" />
+    </div>`)
+
+  const [checkbox, radio, number] = document.querySelectorAll(
+    '[data-testid="element"]',
+  )
+
+  await userEvent.tab()
+
+  expect(radio).toHaveFocus()
+
+  await userEvent.tab()
+
+  expect(checkbox).toHaveFocus()
+
+  await userEvent.tab()
+
+  expect(number).toHaveFocus()
+
+  await userEvent.tab()
+
+  expect(radio).toHaveFocus()
+})
+
+test('should respect dom order when tabindex are all the same', async () => {
```

Comment:
The indices in this test are not all the same.

Reply:
cc @benmonro, I think you wrote these tests. Could you take a look?

Reply:
ah yeah wording is a little confusing now that I read it again.  what about this:
```suggestion
test('should respect tab index order, then DOM order', async () => {
```